feat(protocol)!: implement `PrettyPrintRecordKey` as a `Cow` type

RolandSherwin opened the pull request:
- this allows us to hold a reference to `RecordKey` when we just want to log things.
- We're still cloning while printing `KbucketKey`, will be fixed with an update to the libp2p API.

## Description

reviewpad:summary 
- this allows us to hold a reference to RecordKey when we just want to
  log things.

joshuef commented on sn/src/routing/core/msg_handling/mod.rs:
```diff
@@ -133,9 +136,13 @@ impl Core {
+                        | SystemMsg::JoinAsRelocatedRequest(_)
+                        // The below messages can be skipped from AE checking because we might be
+                        // underway with SAP changes already as it only takes `threshold + 1`
+                        // no. of agreement messages and the rest would be triggering redundant AE flows.
```
Hmm, but there's no guarantee they'd be coming from the same viewpoint...

I'm not sure it's valid to skip here.

RolandSherwin committed:
feat(protocol)!: implement `PrettyPrintRecordKey` as a `Cow` type
- this allows us to hold a reference to RecordKey when we just want to
  log things.

RolandSherwin committed:
feat(protocol): remove allocation inside `PrettyPrintRecordKey::Display`

RolandSherwin committed:
fix(protocol): use custom `Display` for `PrettyPrintKBucketKey`
